src/coreclr/nativeaot/System.Private.CoreLib/src/System/Runtime/TypeLoaderExports.cs

```diff
@@ -28,345 +29,248 @@ public static unsafe void ActivatorCreateInstanceAny(ref object ptrToData, IntPt
             //  3- Update ptrToData to point to that newly allocated object
             ptrToData = RuntimeImports.RhNewObject(pEEType);
 
-            Entry entry = LookupInCache(s_cache, pEETypePtr, pEETypePtr);
-            entry ??= CacheMiss(pEETypePtr, pEETypePtr,
-                    (IntPtr context, IntPtr signature, object contextObject, ref IntPtr auxResult) =>
-                    {
-                        IntPtr result = RuntimeAugments.TypeLoaderCallbacks.TryGetDefaultConstructorForType(new RuntimeTypeHandle(new EETypePtr(context)));
-                        if (result == IntPtr.Zero)
-                            result = RuntimeAugments.GetFallbackDefaultConstructor();
-                        return result;
-                    });
-            RawCalliHelper.Call(entry.Result, ptrToData);
+            if (!TryGetFromCache(pEETypePtr, pEETypePtr, out var v))
+            {
+                v = CacheMiss(pEETypePtr, pEETypePtr,
+                        (IntPtr context, IntPtr signature, object contextObject, ref IntPtr auxResult) =>
+                        {
+                            IntPtr result = RuntimeAugments.TypeLoaderCallbacks.TryGetDefaultConstructorForType(new RuntimeTypeHandle(new EETypePtr(context)));
+                            if (result == IntPtr.Zero)
+                                result = RuntimeAugments.GetFallbackDefaultConstructor();
+                            return result;
+                        });
+            }
+
+            RawCalliHelper.Call(v._result, ptrToData);
         }
 
         //
         // Generic lookup cache
         //
 
-        private class Entry
+        private struct Key : IEquatable<Key>
+        {
+            public IntPtr _context;
+            public IntPtr _signature;
+
+            public Key(nint context, nint signature)
+            {
+                _context = context;
+                _signature = signature;
+            }
+
+            public bool Equals(Key other)
+            {
+                return _context == other._context && _signature == other._signature;
+            }
+
+            public override int GetHashCode()
+            {
+                // pointers will likely match and cancel out in the upper bits
+                // we will rotate context by 16 bit to keep more varying bits in the hash
+                IntPtr context = (IntPtr)BitOperations.RotateLeft((nuint)_context, 16);
```

Comment:
Casing rotates by half of the pointer size. Should we do the same here?

Reply:
Casting basically folds two pointers into a 64bit hash value and then multiplies with golden ratio and picks the topmost bits. That is to ensure that clusters of keys do not map to clusters of buckets. The intermediate unreduced hashcode is native size and we compute it by rotating one part by half the pointer. 

Here we produce a 32bit hash (per `GetHashCode` contract). We are not yet reducing it to the desired length or dispersing among buckets - that will happen inside the cache. Our job here is only to map two pointers to as much unique 32bit number as possible. 
I think most randomness will be in the lower 16 bits, so I rotate by 16 bits. Without rotation it seems more likely to end up with cases where we hash (x0, y0) and (x1, y1) and see the same result.

If I rotate by 32bit, I think the following `nint.GetHashCode()` will undo that by xor-ing upper and lower 32bit and we will get something like carry-less distance between pointers, which is not as random.

Reply:
One other thing to note is that the use of native math in the caches is mostly "because we can". 
I think using 32bit codepath for both 64bit and 32bit case would not make a lot of difference, - not for the sizes of the caches that we expect here anyways, but since we can use 64bit, we use that.

Reply:
Ah ok, I have missed that this goes through a regular GetHashCode.